File: src/lib/calc.ts
export type TokenType = "num" | "op" | "lparen" | "rparen";

export interface Token {
  type: TokenType;
  value: string;
}

function isDigit(c: string): boolean {
  return c >= "0" && c <= "9";
}

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  const s = input.replace(/\s+/g, "");
  let i = 0;
  while (i < s.length) {
    const c = s[i];
    if (c === undefined) break;
    if (isDigit(c) || c === ".") {
      let j = i;
      let dots = 0;
      while (j < s.length) {
        const cj = s[j];
        if (cj === undefined) break;
        if (isDigit(cj)) {
          j++;
        } else if (cj === ".") {
          dots++;
          if (dots > 1) throw new Error("Invalid number");
          j++;
        } else break;
      }
      const value = s.slice(i, j);
      tokens.push({ type: "num", value });
      i = j;
      continue;
    }
    if (c === "+" || c === "-" || c === "*" || c === "/" || c === "%" || c === "(" || c === ")") {
      const type: TokenType = c === "(" ? "lparen" : c === ")" ? "rparen" : "op";
      tokens.push({ type, value: c });
      i++;
      continue;
    }
    throw new Error(`Unexpected character: ${c}`);
  }
  return tokens;
}

const PRECEDENCE: Record<string, number> = { "+": 1, "-": 1, "*": 2, "/": 2, "%": 2 };

function isOpToken(t: Token): boolean {
  return t.type === "op";
}

export function toRPN(tokens: Token[]): Token[] {
  const out: Token[] = [];
  const ops: Token[] = [];
  let prev: Token | null = null;
  for (const t of tokens) {
    if (t.type === "num") {
      out.push(t);
    } else if (t.type === "op") {
      const isUnary = prev === null || prev.type === "op" || prev.type === "lparen";
      if (isUnary && t.value === "-") {
        out.push({ type: "num", value: "0" });
      }
      while (ops.length > 0) {
        const top = ops[ops.length - 1];
        if (
          top &&
          isOpToken(top) &&
          PRECEDENCE[top.value] !== undefined &&
          PRECEDENCE[t.value] !== undefined &&
          PRECEDENCE[top.value] >= PRECEDENCE[t.value]
        ) {
          const popped = ops.pop();
          if (popped) out.push(popped);
        } else break;
      }
      ops.push(t);
    } else if (t.type === "lparen") {
      ops.push(t);
    } else if (t.type === "rparen") {
      let found = false;
      while (ops.length > 0) {
        const top = ops.pop();
        if (!top) break;
        if (top.type === "lparen") {
          found = true;
          break;
        }
        out.push(top);
      }
      if (!found) throw new Error("Mismatched parentheses");
    }
    prev = t;
  }
  while (ops.length > 0) {
    const top = ops.pop();
    if (!top) break;
    if (top.type === "lparen" || top.type === "rparen") throw new Error("Mismatched parentheses");
    out.push(top);
  }
  return out;
}

export function evaluate(input: string): number {
  if (!input.trim()) return 0;
  const tokens = tokenize(input);
  const rpn = toRPN(tokens);
  const stack: number[] = [];
  for (const t of rpn) {
    if (t.type === "num") {
      stack.push(Number(t.value));
    } else if (t.type === "op") {
      const b = stack.pop();
      const a = stack.pop();
      if (a === undefined || b === undefined) throw new Error("Invalid expression");
      switch (t.value) {
        case "+":
          stack.push(a + b);
          break;
        case "-":
          stack.push(a - b);
          break;
        case "*":
          stack.push(a * b);
          break;
        case "/":
          if (b === 0) throw new Error("Division by zero");
          stack.push(a / b);
          break;
        case "%":
          if (b === 0) throw new Error("Division by zero");
          stack.push(a % b);
          break;
        default:
          throw new Error(`Unknown op: ${t.value}`);
      }
    }
  }
  if (stack.length !== 1) throw new Error("Invalid expression");
  const result = stack[0];
  if (result === undefined || !Number.isFinite(result)) throw new Error("Math error");
  return Math.round(result * 1e12) / 1e12;
}

export function formatResult(n: number): string {
  if (!Number.isFinite(n)) return "Error";
  if (Math.abs(n) < 1e-10) return "0";
  const abs = Math.abs(n);
  if (abs >= 1e16 || (abs > 0 && abs < 1e-6)) {
    return n.toExponential(6).replace(/\.?0+e/, "e");
  }
  const fixed = n.toFixed(10);
  return Number(fixed).toString();
}

export function prettyExpr(expr: string): string {
  return expr.replace(/\*/g, "×").replace(/\//g, "÷").replace(/-/g, "−");
}

File: src-tauri/src/calc.rs
#[derive(Debug, Clone, PartialEq)]
enum TokType {
	Num,
	Op,
	LParen,
	RParen,
}

#[derive(Debug, Clone)]
struct Token {
	kind: TokType,
	value: String,
}

fn is_digit(c: char) -> bool {
	c >= '0' && c <= '9'
}

fn tokenize(input: &str) -> Result<Vec<Token>, String> {
	let s: String = input.split_whitespace().collect();
	let bytes = s.as_bytes();
	let mut tokens = Vec::new();
	let mut i = 0;
	while i < bytes.len() {
		let c = bytes[i] as char;
		if is_digit(c) || c == '.' {
			let mut j = i;
			let mut dots = 0;
			while j < bytes.len() {
				let cj = bytes[j] as char;
				if is_digit(cj) {
					j += 1;
				} else if cj == '.' {
					dots += 1;
					if dots > 1 {
						return Err("Invalid number".into());
					}
					j += 1;
				} else {
					break;
				}
			}
			if j == i {
				return Err("Invalid number".into());
			}
			tokens.push(Token {
				kind: TokType::Num,
				value: s[i..j].to_string(),
			});
			i = j;
			continue;
		}
		match c {
			'+' | '-' | '*' | '/' | '%' | '(' | ')' => {
				let kind = if c == '(' {
					TokType::LParen
				} else if c == ')' {
					TokType::RParen
				} else {
					TokType::Op
				};
				tokens.push(Token {
					kind,
					value: c.to_string(),
				});
				i += 1;
			}
			_ => return Err(format!("Unexpected character: {}", c)),
		}
	}
	Ok(tokens)
}

fn precedence(op: &str) -> i32 {
	match op {
		"+" | "-" => 1,
		"*" | "/" | "%" => 2,
		_ => 0,
	}
}

fn to_rpn(tokens: Vec<Token>) -> Result<Vec<Token>, String> {
	let mut out: Vec<Token> = Vec::new();
	let mut ops: Vec<Token> = Vec::new();
	let mut prev: Option<&Token> = None;
	for t in tokens.iter() {
		match t.kind {
			TokType::Num => out.push(t.clone()),
			TokType::Op => {
				let is_unary = match prev {
					None => true,
					Some(p) => p.kind == TokType::Op || p.kind == TokType::LParen,
				};
				if is_unary && t.value == "-" {
					out.push(Token {
						kind: TokType::Num,
						value: "0".into(),
					});
				}
				while let Some(top) = ops.last() {
					if top.kind == TokType::Op
						&& precedence(&top.value) != 0
						&& precedence(&t.value) != 0
						&& precedence(&top.value) >= precedence(&t.value)
					{
						out.push(ops.pop().unwrap());
					} else {
						break;
					}
				}
				ops.push(t.clone());
			}
			TokType::LParen => ops.push(t.clone()),
			TokType::RParen => {
				let mut found = false;
				while let Some(top) = ops.pop() {
					if top.kind == TokType::LParen {
						found = true;
						break;
					}
					out.push(top);
				}
				if !found {
					return Err("Mismatched parentheses".into());
				}
			}
		}
		prev = Some(t);
	}
	while let Some(top) = ops.pop() {
		if top.kind == TokType::LParen || top.kind == TokType::RParen {
			return Err("Mismatched parentheses".into());
		}
		out.push(top);
	}
	Ok(out)
}

fn eval_rpn(rpn: &[Token]) -> Result<f64, String> {
	let mut stack: Vec<f64> = Vec::new();
	for t in rpn {
		match t.kind {
			TokType::Num => {
				let v: f64 = t.value.parse().map_err(|_| "Invalid number".to_string())?;
				stack.push(v);
			}
			TokType::Op => {
				let b = stack.pop().ok_or("Invalid expression")?;
				let a = stack.pop().ok_or("Invalid expression")?;
				let r = match t.value.as_str() {
					"+" => a + b,
					"-" => a - b,
					"*" => a * b,
					"/" => {
						if b == 0.0 {
							return Err("Division by zero".into());
						}
						a / b
					}
					"%" => {
						if b == 0.0 {
							return Err("Division by zero".into());
						}
						a % b
					}
					_ => return Err(format!("Unknown op: {}", t.value)),
				};
				stack.push(r);
			}
			_ => return Err("Invalid token in RPN".into()),
		}
	}
	if stack.len() != 1 {
		return Err("Invalid expression".into());
	}
	let result = stack.pop().unwrap();
	if !result.is_finite() {
		return Err("Math error".into());
	}
	Ok((result * 1e12).round() / 1e12)
}

pub fn evaluate(input: &str) -> Result<f64, String> {
	let trimmed = input.trim();
	if trimmed.is_empty() {
		return Ok(0.0);
	}
	let tokens = tokenize(trimmed)?;
	let rpn = to_rpn(tokens)?;
	eval_rpn(&rpn)
}

pub fn format_result(n: f64) -> String {
	if !n.is_finite() {
		return "Error".into();
	}
	if n.abs() < 1e-10 {
		return "0".into();
	}
	let abs = n.abs();
	if abs >= 1e16 || abs < 1e-6 {
		let s = format!("{:.6e}", n);
		return s.replace("e", "e");
	}
	let fixed = format!("{:.10}", n);
	let parsed: f64 = fixed.parse().unwrap_or(n);
	parsed.to_string()
}

#[tauri::command]
pub fn eval_expression(expr: String) -> Result<String, String> {
	let n = evaluate(&expr)?;
	Ok(format_result(n))
}
